### ISSUE TITLE
Race condition (or property caching problem) resizing ext4 partition

### ISSUE THREAD

**Issue by aruiz** [storaged-project/udisks#572]
I have the following dbus code to resize a filesystem and its partition:
```
def unmount_and_resize (partition):
    """ Unmounts and resizes /boot to 512Mb less at the end, returns offset and new size """
    proxy = dbus.SystemBus().get_object("org.freedesktop.UDisks2",
            "/org/freedesktop/UDisks2/block_devices/" + partition.split("/")[-1])
    props_proxy = dbus.Interface(proxy, "org.freedesktop.DBus.Properties")
    block_proxy =  dbus.Interface(proxy, "org.freedesktop.UDisks2.Block")
    fs_proxy = dbus.Interface(proxy, "org.freedesktop.UDisks2.Filesystem")
    part_proxy = dbus.Interface(proxy, "org.freedesktop.UDisks2.Partition")

    fstype = props_proxy.Get("org.freedesktop.UDisks2.Block", "IdType")
    offset = props_proxy.Get("org.freedesktop.UDisks2.Partition", "Offset")
    size   = props_proxy.Get("org.freedesktop.UDisks2.Filesystem", "Size")
    if fstype != 'ext4':
        raise Exception("We can only resize ext4 /boot partitions")
    fs_proxy.Unmount({})
    fs_proxy.Repair({})
    fs_proxy.Resize(size - 524288000, {})
    block_proxy.Rescan({})
    new_size = props_proxy.Get("org.freedesktop.UDisks2.Filesystem", "Size")

    part_proxy.Resize(new_size, {})
    block_proxy.Rescan({})
    new_part_size = props_proxy.Get("org.freedesktop.UDisks2.Partition", "Size")
    new_part_offset = props_proxy.Get("org.freedesktop.UDisks2.Partition", "Offset")
    print(size, new_size, new_part_size)
    return (new_part_offset, new_part_size)
```

on print() I get the following sizes for the old filesystem size, the new filesystem size and the new partition size:
```
(dbus.UInt64(1072693248L, variant_level=1), dbus.UInt64(1072693248L, variant_level=1), dbus.UInt64(1072693248L, variant_level=1))
```
The second size value should be 500MB smaller than it shows, which causes a problem in resizing the partition. I could blindly resize the partition assuming the same but I wanted to get confirmation from the API on what the new partition size was in case it alinged it somehow so that I made sure that the new partition size was large enough to fit the new filesystem.

**Comment by pothos** [comment]
This means a waiter like https://github.com/storaged-project/udisks/pull/574 is needed in udiskslinuxfilesystem.c `handle_resize` which looks for a change in the size property. But if the filesystem size did not change, then the call will need to hit the timeout (not so dramatic but I think bad UX if more than 5 seconds). For partitions this was avoided by querying the new blocksize via ioctl and then waiting for exactly this value. This could be done here as well by moving the code of 6291c2d16f45f8c6e9ecbc891e24d218a6e50aaa that queries the size from the disk with libblockdev to a common function and use it here.

The recently added filesystem size property is only set for ext4 and xfs and otherwise stays 0.
Therefore all this should only be done if it initially was not 0.

**Comment by aruiz** [comment]
> The recently added filesystem size property is only set for ext4 and xfs and otherwise stays 0.
> Therefore all this should only be done if it initially was not 0.

So far I've been only dealing with ext4 and xfs. By the way, wouldn't it be better to raise an exception with proper error information instead of returning 0?

**Comment by vojtechtrefny** [comment]
These race conditions (functions end before property update) unfortunately exist for all udisks functions. Properties are updated after we get a udev change event so it can take quite a long time before the property value is "right".
Some functions currently wait for changes, but usually only for changes that add or remove objects (e.g. `Unlock`/`Lock` functions wait for the cleartext device to (dis)appear) not for a property update.
So basically we need to add waiting functions to all udisks functions that change properties. Something similar what @pothos did here (thank you).

**Comment by vpodzime** [comment]
> So basically we need to add waiting functions to all udisks functions that change properties. Something similar what @pothos did here (thank you).

Yes, but I'm afraid we need to do this in a more systematic way.
